includes/annotationcommand_adapter.inc

```diff
@@ -0,0 +1,101 @@
+<?php
+
+/**
+ * @file
+ * annotationcommand_adapter.inc
+ */
+
+use Consolidation\AnnotatedCommand\CommandFileDiscovery;
+use Consolidation\AnnotatedCommand\AnnotatedCommandFactory;
+
+/**
+ * TODO: maybe configure and cache the CommandFileDiscovery class?
+ */
+function annotationcommand_adapter_get_discovery() {
+  return new CommandFileDiscovery();
+}
+
+/**
+ * Return the cached commands built by annotationcommand_adapter_discover.
+ * @see drush_get_commands()
+ */
+function annotationcommand_adapter_commands() {
+  $commands = drush_get_context('DRUSH_ANNOTATED_COMMANDS');
+  return $commands;
+}
+
+/**
+ * Search for annotation commands at the provided search path.
+ * @see _drush_find_commandfiles()
+ */
+function annotationcommand_adapter_discover($searchpath, $phase = false, $phase_max = false) {
+  if (empty($searchpath)) {
+    return;
+  }
+  // Check to see if this is the Drush searchpath for instances where we are
+  // NOT going to do a full bootstrap (e.g. when running a help command)
+  if (($phase == DRUSH_BOOTSTRAP_DRUPAL_SITE) && ($phase_max < DRUSH_BOOTSTRAP_DRUPAL_FULL)) {
```

Comment:
I am reluctant to rework how bootstrapping works in the Drush 8 branch; that is what would be necessary to centralize it. Move the code here that finds all of the specific module folders into the bootstrap code. This is not necessary for the existing use case, which searches deeply for .drush.inc files.